### DISCUSSION TITLE
[FIXED] Resource in Dictionary getting garbage collected?

### DISCUSSION THREAD

**Discussion by mdevnull** [quaadgras/graphics.gd#45] [Is this a bug?]
I suspect that godot frees the ImageTexture immediately but I am unsure why.

Given a node struct likee this:
```go
type GodotTwitch struct {
  classdb.Extension[GodotTwitch, Node.Instance]

  EmoteTextureCache Dictionary.Map[string, ImageTexture.Instance] `gd:"emote_texture_cache"`
}
```

And with `h` being the pointer to an instance of thee node above:
```go
path := fmt.Sprintf("user://emotes/%s.png", emoteID.String())
if FileAccess.FileExists() {
  emoteImage := Image.New()

  errInt := emoteImage.Load(path)
  if errInt != 0 {
    Engine.Raise(fmt.Errorf("error loading emote png as image: %w", errInt))
    return
   }

  newTexture := ImageTexture.CreateFromImage(emoteImage)
  h.EmoteTextureCache.SetIndex(emoteID, newTexture)
}
```

After this is called I can see in the Godot editor that the key exists but it points to a <null>
![image](https://github.com/user-attachments/assets/2d2c80c5-e8be-4846-8c0e-2d98267cd363)

**Comment by Splizard** [comment]
I'm going to have a play with Dictionaries and see what is happening here. `ImageTexture` is a resource and should be reference counted, so I'll need to figure out what's happened. I suspect the issue is related to converting the texture into a variant `variant.New(newTexture)`.

On the topic of dictionaries, do you think it would be useful to have a type-safe dictionary type? ie a `Dictionary.Map[K, V]`

**Reply by mdevnull** [reply]
I would love to have a type safe dictionary. I guess as long as it is an option to use `gd.Variant` as the value type no flexibility would be lost.

**Reply by Splizard** [reply]
@devnull-twitch have added a type safe map, and resources should be appropriately referenced in my latest commit so the issue should be fixed now!

**Reply by mdevnull** [reply]
Thanks for the super fix addition!! 🤩 

Should this be usable as a direct property type or should the exported property always be the underlying Dictionary? If I use the Map right now as a property it does not work

```
panic: gdextension.RegisterClass: unsupported property type gd.Map[grow.graphics/gd/internal.String,grow.graphics/gd/internal/classdb.ImageTexture]

goroutine 17 [running, locked to thread]:
grow.graphics/gd.variantTypeOf({0x7fff38a5eeb0, 0x7fff3863f480})
        C:/Users/***/go/pkg/mod/grow.graphics/gd@v0.0.0-20240712101540-523868280049/register_properties.go:160 +0x1345
```

**Reply by Splizard** [reply]
@devnull-twitch ah yes, I have just pushed a fix for this.
